044-cluster-wide-volume-usage-quota-management.md

```diff
@@ -0,0 +1,253 @@
+
+Setting `client.quota.callback.static.storage.hard` or `client.quota.callback.static.storage.soft` would be incompatible
+with the cluster sourced volumes and fail configuration of the plugin.
+
+The admin client will re-use the broker SSL configuration
+
+##### Limit Type Configuration
+
+When operating with cluster sourced volume data the existing limit types applied to the aggregate
+consumed bytes of all volumes would be meaningless. Nodes enter and exit the active set as part of normal operation.
+So we propose that the existing limiting method should be incompatible with cluster sourced volumes.
+
+Instead, we propose introducing new limit types applied on a per-volume basis. Meaning that there is a single value
+for each limit which we test against each volume. i.e. we do not support limiting based on a specific volume.
+
+We retain the existing concept of a soft limit (where throttling begins) and a hard limit (where we effectively
+stop message production) for all limit types.
+
+The only requirement is a single hard-limit configuration, if no soft-limit is configured we will infer that soft-limit==hard-limit.
+
+The different limit types can be combined, so you could combine a consumedBytesExceeds soft limit with a usableRatioBelow
+hard limit.
+
+Defining multiple soft limits or multiple hard limits would be an invalid state.
+
+The limits we want are:
+1. throttle if consumedBytes is greater-than-or-equal-to some threshold for any volume
+2. throttle if usableBytes is less-than-or-equal-to some threshold for any volume
+3. throttle if the ratio of usableBytes to capacityBytes is less-than-or-equal-to some threshold for any volume
+
+Using configuration properties like:
+
+1. `client.quota.callback.static.storage.perVolumeLimit.consumedBytesExceeds.soft=1000`
+2. `client.quota.callback.static.storage.perVolumeLimit.consumedBytesExceeds.hard=5000`
+3. `client.quota.callback.static.storage.perVolumeLimit.usableBytesBelow.soft=5000`
+4. `client.quota.callback.static.storage.perVolumeLimit.usableBytesBelow.hard=1000`
+5. `client.quota.callback.static.storage.perVolumeLimit.usableRatioBelow.soft=0.05`
+6. `client.quota.callback.static.storage.perVolumeLimit.usableRatioBelow.hard=0.01`
+
+`usableRatioBelow` value can be in the range (0.0, 1.0)
+
+#### Fallback Throttle Factor
+
+We are going to depend on using the admin client to get volume information. This brings all the baggage of making a
+connection and dealing with possible failures. Also, we can potentially have an inconsistent view of the world between
+determining the active broker set and asking for a description of all log dirs. So we need to react somehow to cases
+where we cannot get the volume data for all active brokers.
+
+We propose introducing a configurable **fallback throttle factor** to be applied in situations where we don't have enough
+information to act. With a default value of 0.0 ie effectively stopping all production to the cluster. This would allow
```

Comment:
I probably don't agree with this default and would say it should be 1. But I guess it should be discussed as not everyone might share my view ;-).

Reply:
Yeah. The default factor is a fairly opinionated choice. 

My view is that the plug-in should **default** to the safer choice, that is fencing the cluster to prevent a broker running out of disk and all the painful choices that implies. As deploying the quota limiting plug-in explicitly says users want to protect against that scenario. 

However users should be free to accept that risk.

Reply:
I guess we both want the safer option to be default ... but do not agree what the safer option is. One important point for me is that your default deviates from the broker's default state and therefore should be done intentionally by the user and not just because it is default.

Reply:
I think a default of 0 is too pessimistic. By virtue of the fact that this plugin is _being used_ I think we can assume that the users are trying to manage disk space and so it's reasonable to assume they've got the alerting set up etc. Therefore running out of storage is now an extra rare situation to be in, so trading availability for protection against a very rare event seems over cautious.